### PR TITLE
Add support for `torch.Generator` type in TorchScript

### PR DESCRIPTION
**Pull request by antoniojkim**
- Add support for `torch.Generator` type in TorchScript
- Add `generator` args to all `torch.nn.init` functions that call `uniform_` or `normal_`
- Add support for `torch.Generator` in LTC's TorchScript backend (CC: @wconstab)

CC: @eellison @davidberard98 @GlebKazantaev @behzad-a

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
@davidberard98 are you the right person to review this?

**Comment by davidberard98** [comment]
Yeah, can review. But the failures look related, will wait for @antoniojkim to take a look at those first (or ping me if you want help looking at the failures)

**Comment by Skylion007** [comment]
@davidberard98 the lint failure is definitely related?

**Comment by davidberard98** [comment]
linter, and also the test failures look related

**Comment by antoniojkim** [comment]
@albanD Now that this CI appears to be passing for this PR, I wanted to ask if its possible to include the changes from this PR into the 2.1.1 release. We need this feature available in a stable release ideally by the end of the year and I'm assuming that 2.2 isn't happening in 2023.

**Comment by antoniojkim** [comment]
@davidberard98 a gentle reminder to please review when you can

@albanD any thoughts on my above comment?

**Comment by albanD** [comment]
> I wanted to ask if its possible to include the changes from this PR into the 2.1.1 release.

I'm afraid that patch releases are for critical bug-fix only and don't contain any feature work.
2.2 is planned for end of January right now. Hopefully that is soon enough for you. Otherwise, you can always use nightly builds to get this sooner!

File: torch/csrc/lazy/python/init.cpp
Comment (davidberard98) [inline torch/csrc/lazy/python/init.cpp:310]: should this be removed?

**Review by davidberard98 — COMMENTED** [review]
generally lgtm, but can you add more tests? would be great if you can put some in `test/jit` and add a few common/relevant scenarios:
* jit.trace
* jit.script
* using some of the init functions you modified in this PR
* torch.jit.save & torch.jit.load

File: torch/nn/init.py
Comment (wconstab) [inline torch/nn/init.py:152]: @albanD are you ok with these changes to the torch APIs?

File: torchgen/dest/lazy_ir.py
Comment (wconstab) [inline torchgen/dest/lazy_ir.py:291]: can you refresh me on how this works in generated code?

is a Generator object going to be an 'input' to the graph, or a part of the graph?

if it's part of the graph then we should be sure that the lazy graph is recompiling every time you use a new generator object.  Is that the way it works?  Would be good to expand the tests to demonstrate this

Comment (antoniojkim) [reply torchgen/dest/lazy_ir.py:291]: > is a Generator object going to be an 'input' to the graph, or a part of the graph?

It is a part of the graph.

> if it's part of the graph then we should be sure that the lazy graph is recompiling every time you use a new generator object. Is that the way it works?

Yes, the generator state is included in the [hash](https://github.com/antoniojkim/pytorch/blob/antoniojkim/torchscript_generator_support/torch/csrc/lazy/core/hash.h#L152-L154). So, if a different generator with a different state is used, it will recompile.

Recompilation is most likely fine in this case as we don't anticipate users using anything but the default generator inside a training/inference loop. We only anticipate them using it during initialization for deterministic parallel weight initialization.

> Would be good to expand the tests to demonstrate this

Yep, will do

File: torch/csrc/lazy/python/init.cpp
Comment (antoniojkim) [reply torch/csrc/lazy/python/init.cpp:310]: I added it as a tool to verify that the generator was included as part of the TorchScript graph in the LTC flow. I think its a useful utility to have going forward. Thoughts @wconstab?

File: torch/nn/init.py
Comment (albanD) [reply torch/nn/init.py:152]: Yes we are generally ok to add a `generator=None` argument to every function that consumes randomness to allow the user to specify it in a functional way (required by functorch or other functional randomness API).
Most of our factory functions are updated already.

**Review by wconstab — APPROVED** [review]

**Comment by eellison** [comment]
If we're serializing `Generator` type where we didn't previously, we're going to need to FC gate this for a little while.

**Comment by davidberard98** [comment]
@eellison good point; what is involved there? is it (a) add any runtime support and deserialization support first (b) 2 weeks later add frontend capture & serialization support?

**Comment by eellison** [comment]
@davidberard98 yes I believe so

File: CMakeLists.txt
Comment (albanD) [inline CMakeLists.txt:915]: I guess this was removed by mistake?

File: test/allowlist_for_publicAPI.json
Comment (albanD) [inline test/allowlist_for_publicAPI.json:589]: No new entry should be added to this json file.
You can remove this by either adding the appropriate `__all__` entry in the nn.init file or by using `torch.Generator` there to avoid having to make Generator an exposed variable

**Review by albanD — DISMISSED** [review]
build change and public api check needs fixing.
SGTM otherwise.

I don't think FC will be an issue as this only uses the new API in nn.init functions which are used at Model initialization time and thus very unlikely to get serialized into a TS model. So I would suggest that we can land this as is (and revert in the unlikely event that it does break anything).
WDYT @davidberard98 ?

**Comment by davidberard98** [comment]
@albanD thanks for catching the build/allowlist changes

if we don't have serialization changes I think it's fine - but wondering if we'll end up with changes after adding tests for serialization

File: CMakeLists.txt
Comment (antoniojkim) [reply CMakeLists.txt:915]: yep. Not sure why this was committed. Will revert

File: torchgen/dest/lazy_ir.py
Comment (antoniojkim) [reply torchgen/dest/lazy_ir.py:291]: A test was added to verify this behaviour

**Comment by antoniojkim** [comment]
@davidberard98 any ideas why the lintrunner is failing this error?
```
error: 'omp.h' file not found

  #include <omp.h>
```

I also can't reproduce the failure in [this run](https://github.com/pytorch/pytorch/actions/runs/6500762113/job/17657188316?pr=110413). I tried executing the test command locally and it all passed
```
250 passed, 14 skipped, 4 xfailed in 32.83s
```
Maybe its a different environment is why it works for me but fails in CI? Can you please advise how to continue?

@eellison I also seem to now be failing the [backwards compatibility test suite](https://github.com/pytorch/pytorch/actions/runs/6500762113/job/17657211221?pr=110413). Its because I had to add an override for `aten::manual_seed` here
https://github.com/antoniojkim/pytorch/blob/2cf0871fa56b18ea5c919c8034033a249644d0d9/torch/csrc/jit/runtime/register_prim_ops.cpp#L2515-L2524
But because the previously existing `aten::manual_seed` didn't have any overrides it was failing the mobile test suites because it thinks it is registering `aten::manual_seed` multiple times. This is why I added the `aten::manual_seed.default_` override, but that is causing the backwards compatibility test to fail now. Can you please advise on how to proceed?

File: torch/csrc/jit/runtime/register_prim_ops.cpp
Comment (antoniojkim) [inline torch/csrc/jit/runtime/register_prim_ops.cpp:2524]: @eellison I also seem to now be failing the [backwards compatibility test suite](https://github.com/pytorch/pytorch/actions/runs/6500762113/job/17657211221?pr=110413). Its because I had to add an override for `aten::manual_seed` (as highlighted above). But because the previously existing `aten::manual_seed` didn't have any overrides it was failing the mobile test suites because it thinks it is registering `aten::manual_seed` multiple times. This is why I added the `aten::manual_seed.default_` override, but that is causing the backwards compatibility test to fail now. Can you please advise on how to proceed?

**Review by glebk-cerebras — APPROVED** [review]

File: aten/src/ATen/core/Generator.cpp
Comment (davidberard98) [inline aten/src/ATen/core/Generator.cpp:1]: include not needed?

File: torch/csrc/jit/runtime/register_prim_ops.cpp
Comment (davidberard98) [inline torch/csrc/jit/runtime/register_prim_ops.cpp:2498]: maybe take a `Device?` instead of a `str?` - unless there's precedent somewhere else to use `str` for device

File: torch/csrc/jit/runtime/register_prim_ops.cpp
Comment (davidberard98) [reply torch/csrc/jit/runtime/register_prim_ops.cpp:2524]: is it possible to leave `aten::manual_seed` as is and then also add `aten::manual_seed.generator`? Not sure if it will work but I did see some other examples suggesting this might work - e.g. aten::device

File: torch/csrc/jit/runtime/register_prim_ops.cpp
Comment (davidberard98) [inline torch/csrc/jit/runtime/register_prim_ops.cpp:2517]: I think (a) this modifies the generator and (b) the input generator and output generator reference the same object, right?

If so, I think this should be
```
aten::manual_seed.generator(Generator(a!) self, int seed) -> Generator(a!)
```

The `(a)` means that the two generators are aliasing the same memory, and the `!` means that there is mutation.

Not sure of the semantics of `seed()` and `current_seed()` - you may need to update the schemas of some of the other ops?

File: test/jit/test_generator.py
Comment (davidberard98) [inline test/jit/test_generator.py:96]: so this doesn't actually test generators in torchscript; the `__init__` method doesn't get scripted, only the `forward`. Could you update with an example that uses a generator in the forward method?

Also I imagine things might be more difficult to save/load if you have a `self.generator = Generator()..` - because then you'll have to be able to serialize the Generator itself. Can you add a test with this as well? If it doesn't work I think that might also be okay, but maybe add some notes somewhere clarifying that

File: torch/csrc/jit/ir/constants.cpp
Comment (davidberard98) [inline torch/csrc/jit/ir/constants.cpp:115]: are these the semantics we want, i.e. should we be resetting the seed? Will that produce the same behavior in eager as in torchscript?

i.e., this seems to suggest that on each iteration the generator will be producing the same results; whereas in eager mode I'm not sure whether we would always expect the generator to be reset on each iteration.

I guess part of the problem is that I'm not quite sure where we'd want to `tryInsertConstant` for a generator - do you have an example of where this is getting called?

File: torch/csrc/jit/runtime/register_prim_ops.cpp
Comment (antoniojkim) [reply torch/csrc/jit/runtime/register_prim_ops.cpp:2524]: I tried this, but as I pointed out, it caused the following issue:

> it was failing the mobile test suites because it thinks it is registering aten::manual_seed multiple times. This is why I added the aten::manual_seed.default_ override

File: torch/csrc/jit/runtime/register_prim_ops.cpp
Comment (antoniojkim) [reply torch/csrc/jit/runtime/register_prim_ops.cpp:2498]: I used `str?` as that was what was being used by `aten::device`

File: torch/csrc/jit/ir/constants.cpp
Comment (antoniojkim) [reply torch/csrc/jit/ir/constants.cpp:115]: > are these the semantics we want, i.e. should we be resetting the seed? Will that produce the same behavior in eager as in torchscript?

This is a good point. And in fact, this seed setting logic is not needed if we are doing `torch.jit.script`. I added it because I noticed that `torch.jit.trace` doesn't reset the generator seed every time it is run.

Is it an acceptable limitation for now to say that one _has_ to use `torch.jit.script` if you want to use a generator in TorchScript?

File: torch/csrc/jit/runtime/register_prim_ops.cpp
Comment (davidberard98) [reply torch/csrc/jit/runtime/register_prim_ops.cpp:2498]: I think we should use `Device?` instead - `aten::device` is a special case because its purpose is to convert str -> device; although I'm open to keeping it as `str` if it turns out there's other complications with using `Device`.

**Review by davidberard98 — COMMENTED** [review]
minor comment, otherwise LGTM!

please take a look at the linter issue & the asan failure though (btw you can rebase onto the viable/strict branch instead of main to skip over flaky failures on the main branch) branch)

**Comment by antoniojkim** [comment]
> the asan failure though (btw you can rebase onto the viable/strict branch instead of main to skip over flaky failures on the main branch) branch)

I've rebased onto `viable/strict` but that doesn't seem to have helped. Can you please advise how to proceed?

**Comment by davidberard98** [comment]
@antoniojkim the asan failures do seem legit, they haven't failed on any other PRs; you might need to do some debugging to figure out what caused the failure

instructions on getting a local asan build if you want to try to repro locally: https://github.com/pytorch/pytorch/blob/main/CONTRIBUTING.md#building-pytorch-with-asan; though the instructions might be a bit out-of-date, it seems like they are using LLVM 15 instead of LLVM 8.

alternatively if you don't want to set up asan, you could try using gdb/lldb to understand the code pointer referenced in the stacktrace provided by asan?

**Comment by antoniojkim** [comment]
> instructions on getting a local asan build if you want to try to repro locally: https://github.com/pytorch/pytorch/blob/main/CONTRIBUTING.md#building-pytorch-with-asan; though the instructions might be a bit out-of-date, it seems like they are using LLVM 15 instead of LLVM 8.

I was unable to get an ASAN build working. I'm working in a manylinux environment and thus have been using GCC 10 to build PyTorch this entire time. My development environment seems to be incompatible with LLVM 15 due to various library version incompatibilities.

> alternatively if you don't want to set up asan, you could try using gdb/lldb to understand the code pointer referenced in the stacktrace provided by asan?

I don't really have anything to debug locally. The failing test passes for me
```
$ pytest test/lazy/test_ts_opinfo.py::TestLazyOpInfoCPU::test_correctness_native_dropout_backward_cpu_float32
========================================================= test session starts =========================================================
platform linux -- Python 3.8.13, pytest-7.4.2, pluggy-1.3.0
rootdir: pytorch
configfile: pytest.ini
plugins: hypothesis-6.87.1
collected 1 item

test/lazy/test_ts_opinfo.py .                                                                                                   [100%]

========================================================== 1 passed in 5.31s ==========================================================
```
In fact, all of the `test_ts_opinfo.py` tests pass
```
$ pytest test/lazy/test_ts_opinfo.py
========================================================= test session starts =========================================================
platform linux -- Python 3.8.13, pytest-7.4.2, pluggy-1.3.0
rootdir: /net/antonio-dev/srv/nfs/antonio-data/ws/lazy_tensor_staging/pytorch
configfile: pytest.ini
plugins: hypothesis-6.87.1
collected 268 items

test/lazy/test_ts_opinfo.py s........s................ss............................x.....s...s...............................s [ 36%]
................ss............................x.....s...s................................x......x........ss.................... [ 84%]
.......s..................................                                                                                      [100%]

============================================= 250 passed, 14 skipped, 4 xfailed in 44.61s =============================================
```

@davidberard98 Would you be able to help me and see if you can reproduce this using an ASAN build of PyTorch?

**Comment by davidberard98** [comment]
Sorry, I don't think I'll have bandwidth to debug an asan build myself.

my gdb suggestion was to put a breakpoint at the point shown in the stacktrace of the failing test in the CI asan failure.

**Comment by antoniojkim** [comment]
@albanD Seeing as you requested changes, can you please review once more as well

File: docs/source/conf.py
Comment (albanD) [inline docs/source/conf.py:2154]: This needs to be documented, it is not ok to add anything to this list.

**Review by albanD — CHANGES_REQUESTED** [review]
scratch that, the public api doc was not fixed

File: docs/source/conf.py
Comment (antoniojkim) [reply docs/source/conf.py:2154]: @albanD Much like all of the other `*Type` classes, its a new non user facing type for which docs won't exist. I think its fine to add this class to the `coverage_ignore_classes`. Is there a particular reason why it can't?

Comment (albanD) [reply docs/source/conf.py:2154]: Because we have a strict policy that all public APIs must be documented.

There are quite a few historical ones that are not yet documented but we want to fix these to empty this list.

Comment (antoniojkim) [reply docs/source/conf.py:2154]: Given that there are no other `*Type` classes documented anywhere, its unclear where such documentation should be added. The `GeneratorType` jit type has existed long before this PR. Its just only now part of the public API as there was seemingly never a need for it in the past. But now there is as TorchScript infers its types from the type annotation (see changes to `torch/jit/annotations.py`).

Seeing as none of the other `*Type` classes are documented anywhere either, I think this falls into the "historical" category. When all the jit `*Type` classes are documented, the `GeneratorType` can be documented at the same time.

Comment (albanD) [reply docs/source/conf.py:2154]: I'm fine with moving this type to be private (rename it to _GeneratorType) if we say that we don't want to document it right now.
Then we can move it to a public once we properly document it.

Comment (antoniojkim) [reply docs/source/conf.py]: done

**Review by albanD — APPROVED** [review]
Sounds good to me.
@davidberard98 any last comment before we merge?

**Comment by davidberard98** [comment]
@antoniojkim @albanD I guess I'm still not sure whether we should be skipping the asan failure? IIRC the stack trace _did_ use one of the random ops from `random_ops.cpp` that was removed in this PR; do we have reason to believe that the failures is unrelated to these changes or otherwise not important?

**Comment by antoniojkim** [comment]
> IIRC the stack trace _did_ use one of the random ops from `random_ops.cpp` that was removed in this PR

Did it? I don't recall seeing that. I removed `torch/csrc/lazy/ts_backend/ops/random_ops.{h,cpp}` as nothing in those files was being used even prior to this PR.

> Do we have reason to believe that the failures is unrelated to these changes or otherwise not important?

The test _only_ appears to be failing with ASAN in CI. Locally and I'm guessing in at least one of the other CI suites, it runs fine.

This was a test that was completely xfailed before. So, by skipping it in the ASAN suite, I'm actually adding it to be run in other suites.

**Review by davidberard98 — APPROVED** [review]
awesome - didn't realize it was skipped before.

thanks for the great work on this PR!

**Comment by PaliC** [comment]
@pytorchbot revert -m "breaking internal builds" -c "ghfirst"

The errors are all build errors with the final following error message 

```
[pytorch_dir]/torch/csrc/lazy/python/init.cpp:315:40: error: unknown type name 'TSComputation'; did you mean 'Computation'?
[CONTEXT]     auto ts_computation = dynamic_cast<TSComputation*>(computation.get());
[CONTEXT]                                        ^~~~~~~~~~~~~
[pytorch_dir]/torch/csrc/lazy/python/init.cpp:317:28: error: no member named 'graph' in 'torch::lazy::Computation'
[CONTEXT]     return ts_computation->graph()->toString();
[CONTEXT] 
```

**Comment by antoniojkim** [comment]
@PaliC Can you please confirm whether the latest commit which I just added would fix the internal build issues? (CC: @wconstab)

**Comment by antoniojkim** [comment]
@davidberard98 @PaliC Any updates on whether the internal builds are okay now?

**Comment by antoniojkim** [comment]
@davidberard98 @PaliC Am I good to start merging this PR again?

**Comment by PaliC** [comment]
Yep it looks good to me
@pytorchbot merge

**Comment by huydhn** [comment]
@pytorchbot revert -m "Sorry for reverting your change but it is, unfortunately, still breaking internal builds" -c "ghfirst"

The failure come from an internal meta device that PyTorch supports (not CUDA or MPS) https://www.internalfb.com/sandcastle/workflow/3598376102270459992.

I will re-import this PR after reverting because I think an internal fix might be needed.  cc @davidberard98 Could you help take a look to see how to unblock this?

**Comment by antoniojkim** [comment]
@huydhn @davidberard98 Any updates on this?

File: aten/src/ATen/core/Generator.cpp
Comment (davidberard98) [inline aten/src/ATen/core/Generator.cpp:24]: Can you move this function into `torch/` somewhere? I think this is the source of the build issue.

e.g. for now you can put it in `torch/csrc/jit/runtime/`. (I know ideally we probably would like this to be somewhere in ATen, but for now I think it will be a good way to work around the build issue - then we can follow up other build experts on whether there's a good way to put this in ATen.

**Comment by davidberard98** [comment]
@huydhn I think this fixes the build issue. I'm going to merge it tomorrow unless you have any objections.

**Pull request closed without merging**
